### PR TITLE
feat(Core/DBUpdater): implement db auto update

### LINKED ISSUE #3739: Feature: Automatic DB updates via worldserver [$50]
**Describe your feature request or suggestion in detail**

Basically, the way TC handles DB updates but more progressive instead of having our external tool.

So when worldserver.exe is started it checks the latest SQL file in the src to see if the same revision is in the current DB or if any other file is missing (incase they have been imported manually in weird orders) - if not we automatically import the missing updates.

This would simplify keeping the DB up-to-date and first startup of the core instead of using a *in my opinion* confusing external tool.

Obviously this needs to have an option in config to disable auto updates

<bountysource-plugin>

---
There is a **[$50 open bounty](https://www.bountysource.com/issues/94522364-feature-automatic-db-updates-via-worldserver?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## Changes Proposed:
- The long-awaited automatic database update feature has been added

## Issues Addressed:
- Closes #3739 

## Tests Performed:
-
image
image

## How to Test the Changes:
1. Just change options in config `*DatabaseInfo` to new DB
2. Wait for the startup
3. Done
4. Example https://gist.github.com/Winfidonarleyan/5baa31f53894e30d61e96ac1df2359d5

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/cmake/macros/ConfigureModules.cmake [hunk 1]
```diff
@@ -71,3 +71,18 @@ function(IsDynamicLinkingModulesRequired variable)
...
+    ${CMAKE_SOURCE_DIR}/modules/*)
+
+  foreach(MODULE_DIR ${LOCALE_MODULE_LIST})
+    if(IS_DIRECTORY "${CMAKE_SOURCE_DIR}/modules/${MODULE_DIR}")
+      set(MODULE_LIST__ ${MODULE_LIST__}${MODULE_DIR},)
+    endif()
+  endforeach()
+
+  add_definitions(-DAC_MODULES_LIST=$<1:"${MODULE_LIST__}">)
+endfunction()
```
Comment (Winfidonarleyan) [inline src/cmake/macros/ConfigureModules.cmake:88]: .

File: src/cmake/macros/FindMySQL.cmake [hunk 2]
```diff
@@ -230,7 +276,10 @@ if( MYSQL_LIBRARY )
...
   endif( MYSQL_INCLUDE_DIR )
-  mark_as_advanced( MYSQL_FOUND MYSQL_LIBRARY MYSQL_EXTRA_LIBRARIES MYSQL_INCLUDE_DIR )
+  if( MYSQL_EXECUTABLE )
+    message(STATUS "Found MySQL executable: ${MYSQL_EXECUTABLE}")
+  endif( MYSQL_EXECUTABLE )
+  mark_as_advanced( MYSQL_FOUND MYSQL_LIBRARY MYSQL_EXTRA_LIBRARIES MYSQL_INCLUDE_DIR MYSQL_EXECUTABLE )
 else( MYSQL_LIBRARY )
   message(FATAL_ERROR "Could not find the MySQL libraries! Please install the development libraries and headers")
-endif( MYSQL_LIBRARY )
+endif( MYSQL_LIBRARY )
```
Comment (Winfidonarleyan) [inline src/cmake/macros/FindMySQL.cmake:285]: .

File: src/common/Configuration/BuiltInConfig.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (C) 2016+ AzerothCore <www.azerothcore.org>, released under GNU AGPL v3 license: https://github.com/azerothcore/azerothcore-wotlk/blob/master/LICENSE-AGPL3
+ * Copyright (C) 2021+ WarheadCore <https://github.com/WarheadCore>
+ */
+
```
Comment (Winfidonarleyan) [inline src/common/Configuration/BuiltInConfig.cpp:5]: .

File: src/common/Configuration/BuiltInConfig.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (C) 2016+ AzerothCore <www.azerothcore.org>, released under GNU AGPL v3 license: https://github.com/azerothcore/azerothcore-wotlk/blob/master/LICENSE-AGPL3
+ * Copyright (C) 2021+ WarheadCore <https://github.com/WarheadCore>
+ */
+
```
Comment (Winfidonarleyan) [inline src/common/Configuration/BuiltInConfig.h:5]: .

File: src/common/Utilities/StartProcess.h [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    /// Note that most executables expect it's name as the first argument.
+    AC_COMMON_API std::shared_ptr<AsyncProcessResult> StartAsyncProcess(std::string executable, std::vector<std::string> args,
+            std::string logger, std::string input_file = "",
+            bool secure = false);
+
+    /// Searches for the given executable in the PATH variable
+    /// and returns a non-empty string when it was found.
+    AC_COMMON_API std::string SearchExecutableInPath(std::string const& filename);
+
+} // namespace Warhead
```
Comment (Winfidonarleyan) [inline src/common/Utilities/StartProcess.h:55]: .

**Comment by Winfidonarleyan** [comment]
Ready to testing

**Comment by Krejza9** [comment]
Good job man

File: src/server/authserver/authserver.conf.dist [hunk 6]
```diff
@@ -134,6 +135,29 @@ WrongPass.BanType = 0
 
 WrongPass.Logging = 0
 
+#
+#    SourceDirectory
+#        Description: The path to your WarheadCore source directory.
```
Comment (Krejza9) [inline src/server/authserver/authserver.conf.dist:140]: wrong directory

File: src/server/authserver/authserver.conf.dist [hunk 7]
```diff
@@ -134,6 +135,29 @@ WrongPass.BanType = 0
...
+#                     If the path is left empty, the built-in CMAKE_SOURCE_DIR is used.
+#        Example:     "../WarheadCore"
```
Comment (Krejza9) [inline src/server/authserver/authserver.conf.dist:142]: here too

File: src/cmake/macros/FindMySQL.cmake [hunk 8]
```diff
@@ -222,6 +234,40 @@ else( NOT WIN32 )
...
+      /usr/bin/
+    DOC
+      "path to your mysql binary."
+  )
+endif( UNIX )
+
+if( WIN32 )
+  find_program(MYSQL_EXECUTABLE mysql
+    PATHS
+      "C:/Program Files/MySQL/MySQL Server 5.7/bin"
```
Comment (Kitzunu) [inline src/cmake/macros/FindMySQL.cmake:252]: also need to set for 8.0?

**Comment by Kitzunu** [comment]
Seems to work pretty great, I dropped my world DB and it filled out amazingly =)

One note:
Is there a way to suppress this error as it spams on every file being imported
```
mysql: [Warning] Using a password on the command line interface can be insecure.
```

**Comment by Winfidonarleyan** [comment]
> Seems to work pretty great, I dropped my world DB and it filled out amazingly =)

Thanks for testing.

> One note:
> Is there a way to suppress this error as it spams on every file being imported
> 
> ```
> mysql: [Warning] Using a password on the command line interface can be insecure.
> ```

At this point, I don't know if it's possible to do this

And thanks for your help

File: src/cmake/macros/FindMySQL.cmake (hunk 8 above)
Comment (Winfidonarleyan) [reply src/cmake/macros/FindMySQL.cmake]: Added

**Comment by Winfidonarleyan** [comment]
@Krejza9 thanks for review

**Comment by FrancescoBorzi** [comment]
@Winfidonarleyan @Krejza9 @Kitzunu  

so is this ready to be merged or requires more testing? please update the labels accordingly

**Review by Kitzunu — APPROVED** [review]

**Comment by Kitzunu** [comment]
Poggers

**Comment by Winfidonarleyan** [comment]
Yay, it happened.

**Comment by Bogir** [comment]
On first start, the server overwrote several of my custom NPC (for some mods)
and I did not find in the log of the report which NPC the server updated and what exactly changed.
Now, to fix it, I had to roll back both the core and the base back to March 2021. 

Suggestion: check existing records before adding or before replacing them.  And stop loading server if founded conflicts.

**Comment by FrancescoBorzi** [comment]
> On first start, the server overwrote several of my custom NPC (for some mods)
> and I did not find in the log of the report which NPC the server updated and what exactly changed.
> Now, to fix it, I had to roll back both the core and the base back to March 2021.
> 
> Suggestion: check existing records before adding or before replacing them. Stop loading server if founded conflicts and give the user the ability to ignore or accept changes in the event of a conflict .

That's not a good approach. You should keep your custom scripts in separate SQL files and re-apply them every time you update your server. Make sure your SQL queries are re-executable (so use `DELETE` before `INSERT`).

**Comment by Bogir** [comment]
> 
> That's not a good approach. You should keep your custom scripts in separate SQL files and re-apply them every time you update your server. Make sure your SQL queries are re-executable (so use `DELETE` before `INSERT`).

I don't agree. AutoUpdater must handle it. 
Custom changes haven't SQL files. It was direct fixes/changes in DB.

Another example:
https://discord.com/channels/217589275766685707/284406375495368704/861325133540360193

**Pull request merged**